File: packages/sdk-go/conformance/unsigned_requests_test.go
package conformance

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"testing"

	"github.com/gemini/developer-platform/packages/sdk-go"
	"github.com/gemini/developer-platform/packages/sdk-go/generated/predictions"
)

func TestUnsignedRequests(t *testing.T) {
	manifest := LoadManifest(t)
	var suite FixtureSuite
	for _, candidate := range manifest.Suites {
		if candidate.ID == "http/unsigned-requests" {
			suite = candidate
			break
		}
	}
	if suite.ID == "" {
		t.Fatal("manifest does not contain http/unsigned-requests suite")
	}
	for _, caseID := range suite.Cases {
		t.Run(caseID, func(t *testing.T) {
			var fixture FixtureUnsigned
			if err := json.Unmarshal(LoadCase(t, suite.ID, caseID), &fixture); err != nil {
				t.Fatalf("decoding fixture: %v", err)
			}
			if fixture.ID != suite.ID+"/"+caseID || fixture.Kind != suite.Kind {
				t.Fatalf("fixture identity = (%q, %q), want (%q, %q)", fixture.ID, fixture.Kind, suite.ID+"/"+caseID, suite.Kind)
			}
			testUnsignedRequest(t, fixture)
		})
	}
}

func testUnsignedRequest(t *testing.T, fixture FixtureUnsigned) {
	t.Helper()
	var observed *http.Request
	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		observed = req.Clone(req.Context())
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{}`))
	}))
	defer server.Close()

	client := gemini.NewClient(
		gemini.WithEnvironment(gemini.Sandbox),
		gemini.WithCustomRESTURL(server.URL),
		gemini.WithHTTPClient(server.Client()),
	)
	defer client.Close()

	ctx := context.Background()
	switch fixture.Operation {
	case "marketData.getTicker":
		symbol := requiredStringInput(t, fixture.Input, "symbol")
		if _, err := client.MarketData.GetTicker(ctx, symbol); err != nil {
			t.Fatalf("MarketData.GetTicker failed: %v", err)
		}
	case "marketData.getCurrentOrderBook":
		symbol := requiredStringInput(t, fixture.Input, "symbol")
		limitBids := requiredIntInput(t, fixture.Input, "limit_bids")
		limitAsks := requiredIntInput(t, fixture.Input, "limit_asks")
		if _, err := client.MarketData.GetOrderBook(ctx, symbol, limitBids, limitAsks); err != nil {
			t.Fatalf("MarketData.GetOrderBook failed: %v", err)
		}
	case "predictions.listEvents":
		var statusValues []string
		raw, ok := fixture.Input["status"]
		if !ok {
			t.Fatal("prediction fixture input is missing status")
		}
		if err := json.Unmarshal(raw, &statusValues); err != nil {
			t.Fatalf("decoding status input: %v", err)
		}
		statuses := make([]predictions.MarketStatus, len(statusValues))
		for i, value := range statusValues {
			statuses[i] = predictions.MarketStatus(value)
		}
		params := &predictions.ListEventsParams{Status: &statuses}
		if _, err := client.Predictions.GetEvents(ctx, params); err != nil {
			t.Fatalf("Predictions.GetEvents failed: %v", err)
		}
	default:
		t.Fatalf("unsupported unsigned operation %q", fixture.Operation)
	}
	if observed == nil {
		t.Fatal("service request did not reach httptest handler")
	}
	if observed.Method != fixture.Expect.Method {
		t.Errorf("method = %q, want %q", observed.Method, fixture.Expect.Method)
	}
	if observed.URL.Path != fixture.Expect.Path {
		t.Errorf("path = %q, want %q", observed.URL.Path, fixture.Expect.Path)
	}
	if got, want := observed.URL.Query(), fixture.Expect.Query; !reflect.DeepEqual(map[string][]string(got), want) {
		t.Errorf("query = %v, want %v", got, want)
	}
	for _, name := range fixture.Expect.AuthHeaders {
		if observed.Header.Get(name) == "" {
			t.Errorf("expected non-empty authentication header %q", name)
		}
	}
	for _, name := range []string{"Authorization", "X-GEMINI-APIKEY", "X-GEMINI-NONCE", "X-GEMINI-PAYLOAD", "X-GEMINI-SIGNATURE"} {
		if !contains(fixture.Expect.AuthHeaders, name) && observed.Header.Get(name) != "" {
			t.Errorf("unsigned request unexpectedly included %s=%q", name, observed.Header.Get(name))
		}
	}
}

func requiredStringInput(t *testing.T, input map[string]json.RawMessage, name string) string {
	t.Helper()
	raw, ok := input[name]
	if !ok {
		t.Fatalf("input is missing %s", name)
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil || value == "" {
		t.Fatalf("input %s is not a non-empty string: %s", name, raw)
	}
	return value
}

func requiredIntInput(t *testing.T, input map[string]json.RawMessage, name string) int {
	t.Helper()
	raw, ok := input[name]
	if !ok {
		t.Fatalf("input is missing %s", name)
	}
	var value int
	if err := json.Unmarshal(raw, &value); err != nil {
		t.Fatalf("input %s is not an integer: %s (%v)", name, raw, err)
	}
	return value
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

File: packages/sdk-typescript/src/tests/conformance/unsigned-requests.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { PredictionMarketsRest } from "../../generated/rest.js";
import { MarketDataRest } from "../../generated/market-data/rest.js";
import { HttpTransport, type FetchLike, type HttpMethod } from "../../transport/http.js";
import { type BoundaryRecord } from "../../utils/boundary-value.js";
import { streamingTextResponse } from "../support/http-fixtures.js";
import {
  loadCase,
  loadManifest,
  type FixtureCase,
} from "./support/fixtures.js";

type UnsignedFixture = FixtureCase & {
  operation: "marketData.getTicker" | "marketData.getCurrentOrderBook" | "predictions.listEvents";
  input?: BoundaryRecord;
  expect: {
    method: HttpMethod;
    path: string;
    query?: Record<string, readonly string[]>;
    authHeaders: readonly string[];
  };
};

function fixture(value: FixtureCase): UnsignedFixture {
  return value as unknown as UnsignedFixture;
}

function assertQuery(url: URL, expected: Record<string, readonly string[]> | undefined): void {
  const expectedEntries = Object.entries(expected ?? {});
  const actualKeys = [...new Set([...url.searchParams.keys()])].sort();
  assert.deepEqual(actualKeys, expectedEntries.map(([name]) => name).sort());
  for (const [name, values] of expectedEntries) {
    assert.deepEqual(url.searchParams.getAll(name), [...values], `query ${name}`);
  }
}

async function runCase(value: UnsignedFixture): Promise<void> {
  let captured: { url: string; init: Parameters<FetchLike>[1] } | undefined;
  const transport = new HttpTransport({
    env: "sandbox",
    maxRetries: 0,
    fetchImpl: async (url, init) => {
      captured = { url, init };
      return streamingTextResponse("{}", 200, { get: () => "application/json" });
    },
  });
  const marketData = new MarketDataRest(transport);
  const predictions = new PredictionMarketsRest(transport);
  switch (value.operation) {
    case "marketData.getTicker":
      await marketData.getTicker(value.input as never);
      break;
    case "marketData.getCurrentOrderBook":
      await marketData.getCurrentOrderBook(value.input as never);
      break;
    case "predictions.listEvents":
      await predictions.listEvents(value.input as never);
      break;
    default:
      throw new Error(`unsupported unsigned conformance operation: ${value.operation}`);
  }
  assert.ok(captured, "fixture request did not reach fetch");
  const url = new URL(captured.url);
  assert.equal(captured.init.method, value.expect.method);
  assert.equal(url.pathname, value.expect.path);
  assertQuery(url, value.expect.query);
  assert.equal(captured.init.body, undefined);
  const actualAuthHeaders = Object.keys(captured.init.headers)
    .filter((name) => name.toLowerCase().startsWith("x-gemini-") || name.toLowerCase() === "authorization")
    .map((name) => name.toLowerCase())
    .sort();
  assert.deepEqual(
    actualAuthHeaders,
    value.expect.authHeaders.map((name) => name.toLowerCase()).sort(),
    "authentication headers must match the fixture",
  );
  for (const name of value.expect.authHeaders) {
    const actual = Object.entries(captured.init.headers)
      .find(([header]) => header.toLowerCase() === name.toLowerCase())?.[1];
    assert.notEqual(actual, undefined, `expected auth header ${name}`);
  }
  for (const header of Object.values(captured.init.headers)) {
    assert.notEqual(header, "", "empty header");
  }
}
const manifest = await loadManifest();
const suite = manifest.suites.find((entry) => entry.id === "http/unsigned-requests");
if (!suite) throw new Error("conformance manifest is missing http/unsigned-requests");
for (const caseId of suite.cases) {
  test(`conformance unsigned request: ${caseId}`, async () => {
    await runCase(fixture(await loadCase(suite.id, caseId)));
  });
}
